gonat_linux.go

```diff
@@ -0,0 +1,343 @@
+	if err != nil {
+		return err
+	}
+	ops.Go(func() { s.readFromUpstream(s.tcpSocket) })
+
+	s.udpSocket, err = createSocket(FiveTuple{IPProto: syscall.IPPROTO_UDP, Src: Addr{s.opts.IFAddr, 0}})
+	if err != nil {
+		s.tcpSocket.Close()
+		return err
+	}
+	ops.Go(func() { s.readFromUpstream(s.udpSocket) })
+
+	s.ctrack, err = conntrack.Dial(nil)
+	if err != nil {
+		s.tcpSocket.Close()
+		s.udpSocket.Close()
+		return errors.New("Unable to obtain connection for managing conntrack: %v", err)
+	}
+
+	ops.Go(s.trackStats)
+	ops.Go(s.dispatch)
+	ops.Go(s.writeToDownstream)
+	return s.readFromDownstream()
+}
+
+func (s *server) dispatch() {
+	defer func() {
+		for _, c := range s.connsByDownFT {
+			c.Close()
+			s.deleteConn(c)
+			s.deleteConntrackEntry(c.upFT)
+		}
+		close(s.toDownstream)
+		s.tcpSocket.Close()
+		s.udpSocket.Close()
+		s.ctrack.Close()
+		close(s.closed)
+	}()
+
+	reapTicker := time.NewTicker(1 * time.Second)
+	defer reapTicker.Stop()
+
+	for {
+		select {
+		case pkt := <-s.fromDownstream:
+			s.onPacketFromDownstream(pkt)
+		case pkt := <-s.fromUpstream:
+			s.onPacketFromUpstream(pkt)
+		case c := <-s.closedConns:
+			s.deleteConntrackEntry(c.upFT)
+		case <-reapTicker.C:
+			s.reapIdleConns()
+		case <-s.close:
+			return
+		}
+	}
+}
+
+func (s *server) onPacketFromDownstream(pkt *IPPacket) {
+	switch pkt.IPProto {
+	case syscall.IPPROTO_TCP, syscall.IPPROTO_UDP:
+		s.opts.OnOutbound(pkt)
+		downFT := pkt.FT()
+		c := s.connsByDownFT[downFT]
+
+		if pkt.HasTCPFlag(TCPFlagRST) {
+			if c != nil {
+				c.Close()
+			}
+			return
+		}
+
+		if c == nil {
+			upFT, err := s.assignPort(downFT)
+			if err != nil {
+				log.Errorf("Unable to assign port, dropping packet %v: %v", downFT, err)
+				s.dropPacket(pkt)
+				return
+			}
+			c, err = s.newConn(downFT, upFT)
+			if err != nil {
+				log.Errorf("Unable to create connection, dropping packet %v: %v", downFT, err)
+				s.dropPacket(pkt)
+				return
+			}
+			s.connsByDownFT[downFT] = c
+			s.connsByUpFT[upFT] = c
+			c.markActive()
+			s.addConn(pkt.IPProto)
+		}
+		select {
+		case c.toUpstream <- pkt:
+			log.Tracef("Transmit --  %v -> %v", c.downFT, c.upFT)
+			s.acceptedPacket()
+		default:
+			// don't block if we're stalled writing upstream
+			log.Tracef("Stalled writing packet %v upstream", downFT)
+			s.dropPacket(pkt)
+		}
+	default:
+		log.Tracef("Unknown IP protocol, ignoring packet %v: %v", pkt.FT(), pkt.IPProto)
+		s.dropPacket(pkt)
+	}
+}
+
+func (s *server) onPacketFromUpstream(pkt *IPPacket) {
+	upFT := pkt.FT().Reversed()
+	c := s.connsByUpFT[upFT]
+	if c == nil {
+		log.Tracef("Dropping packet for unknown upstream FT: %v", upFT)
+		s.dropPacket(pkt)
+		return
+	}
+
+	pkt.SetDest(c.downFT.Src)
+	c.s.opts.OnInbound(pkt, c.downFT)
+	pkt.recalcChecksum()
+	c.s.acceptedPacket()
+	c.markActive()
+	log.Tracef("Transmit -- %v <- %v", c.downFT, c.upFT)
+	c.s.toDownstream <- pkt
```

Comment:
I find it odd that we're using the conn's server field (which should == s) instead of s directly.  Is there a reason for that?

I actually find it odd that conn has a reference back to the server.  It seems that it only needs this to register when it has closed.  It could just get the closedConns channel or a function to call when it closes.

Reply:
Yes, I'll change it in this spot. conn does need a reference to server inside of conn.writeToUpstream() to get the buffer pool. We could give conn just the buffer pool, but I don't think it makes a big difference either way.

Reply:
Ah right.  Okay.